### PR TITLE
Implement STDOUT Report Functionality in s3_to_sfts.py

### PR DESCRIPTION
**Pull request by lclarko**

### CODE REVIEW

File: sfts/s3_to_sfts.py [hunk 1]
```diff
@@ -161,19 +225,24 @@ def is_processed():
 try:
     xfer_jar = f"{xfer_path}/xfer.jar"
     jna_jar = f"{xfer_path}/jna.jar"
-    print(("trying to call subprocess:\nxfer.jar: "
+    logger.info(("trying to call subprocess:\nxfer.jar: "
            f"{xfer_jar}\njna.jar : {jna_jar}"))
     subprocess.check_call(
```
Comment (mark-walle) [inline sfts/s3_to_sfts.py:230]: We may want to change this `check_call` to `check_output` https://docs.python.org/3/library/subprocess.html#subprocess.check_output.

Capture the output somehow, like `output = subprocess.check_output(...)`, then shortly after you could log it: `logger.info(output)`. That ought to keep the cron report clean, while preserving any messaging we might want to have for later review.

The exception raised by `check_output` is the same `CalledProcessError` one, so there shouldn't be need to alter the `try ... except` block.

File: sfts/s3_to_sfts.py [hunk 2]
```diff
@@ -161,19 +225,24 @@ def is_processed():
...
         ["java", "-classpath", f"{xfer_jar}:{jna_jar}",
          "xfer",
          f"-user:{sfts_user}",
          f"-password:{sfts_pass}",
          f"-s:{sfts_conf}",
-         "filetransfer.gov.bc.ca"])
+         "filetransfer.gov.bc.ca"],
+        stdout=subprocess.DEVNULL,
+        stderr=subprocess.DEVNULL)
```
Comment (mark-walle) [inline sfts/s3_to_sfts.py:238]: From https://docs.python.org/3/library/subprocess.html#subprocess.DEVNULL

> Special value that can be used as the stdin, stdout or stderr argument to Popen and indicates that the special file os.devnull will be used.

If the output is to the null device, I think we can't ever really review what the output was. Correct me if I'm wrong.

We want to keep the log of the output of the subprocess call somewhere for review.

**Review by mark-walle — CHANGES_REQUESTED** [review]
@lclarko I've submitted some inline comments for your consideration.

File: sfts/s3_to_sfts.py [hunk 3]
```diff
@@ -110,6 +122,55 @@ def is_processed():
     return False
 
 
+# Will run at end of script to print out accumulated report_stats
+def report(data):
+    '''reports out cumulative script events'''
+    print(f'Report: {__file__}\n')
+    print(f'Config: {configfile}')
+    if data['objects_failed_to_sfts'] or data['objects_not_processed']:
```
Comment (lclarko) [inline sfts/s3_to_sfts.py:130]: This is my favourite part

File: sfts/s3_to_sfts.py (hunk 1 above)
Comment (mark-walle) [reply sfts/s3_to_sfts.py]: You've switched to `check_output`. Resolving this.

File: sfts/s3_to_sfts.py [hunk 4]
```diff
@@ -161,19 +225,23 @@ def is_processed():
 try:
     xfer_jar = f"{xfer_path}/xfer.jar"
     jna_jar = f"{xfer_path}/jna.jar"
-    print(("trying to call subprocess:\nxfer.jar: "
+    logger.info(("trying to call subprocess:\nxfer.jar: "
            f"{xfer_jar}\njna.jar : {jna_jar}"))
-    subprocess.check_call(
+    output = subprocess.check_output(
```
Comment (mark-walle) [inline sfts/s3_to_sfts.py:230]: @lclarko `output` is a byte literal (see: https://docs.python.org/3/library/stdtypes.html#binary-sequence-types-bytes-bytearray-memoryview).

This results in logs that will be difficult to read at a glance:
```
INFO:__main__:2021-05-13 11:03:05,994:b'User GDXSFTSDEV signed on\n--> cd GCPE/Shared/gdxtest\n--> put ./tmp/test_pmrp_20200228_20200327_20200330T183405_part000.csv\nUpload complete. File ID 767097304\n125126 bytes sent in 1.94 seconds at 62.99 kB/s\n--> quit\n'
```
You'll see there that the output is prepended with `b`, confirming a byte literal.

There are flags for `subprocess.check_output` that should resolve this issue.

From https://docs.python.org/3/library/subprocess.html#frequently-used-arguments
> If text mode is not used, stdin, stdout and stderr will be opened as binary streams. No encoding or line ending conversion is performed.

Please try setting the `text` flag true. You might also need to set the `encoding` flag; I'm not sure.

File: sfts/s3_to_sfts.py [hunk 5]
```diff
@@ -161,19 +225,23 @@ def is_processed():
...
         ["java", "-classpath", f"{xfer_jar}:{jna_jar}",
          "xfer",
          f"-user:{sfts_user}",
          f"-password:{sfts_pass}",
          f"-s:{sfts_conf}",
          "filetransfer.gov.bc.ca"])
     xfer_proc = True
+    logger.info(output)
```
Comment (mark-walle) [inline sfts/s3_to_sfts.py:238]: If you choose not to use an encoding flag on `subprocess.check_output` above, then you may alternatively decode this here.

Something like: `logger.info(output.decode('utf-8'))` would be a place to start.

**Review by mark-walle — CHANGES_REQUESTED** [review]

**Review by mark-walle — CHANGES_REQUESTED** [review]
@lclarko you have modified the `Pipfile`. This means there would have been a new `Pipfile.lock` generated. Please commit that so in Production we can run `pipenv install --ignore-pipfile` and not cause a lockfile drift in Prod from what's in the main branch.

File: sfts/s3_to_sfts.py (hunk 4 above)
Comment (mark-walle) [reply sfts/s3_to_sfts.py:230]: decoding on log output instead.

**Review by mark-walle — APPROVED** [review]

**Pull request merged**